packages/theming-core/src/generics/themeRegistry.ts

```diff
@@ -0,0 +1,247 @@
+  const graph: IThemeGraph<ITheme, IThemeDefinition> = {
+    _platform: { resolved: platformDefaults },
+    default: { parent: _platformKey }
+  };
+  return {
+    graph,
+    registerTheme: (definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>, name?: string, parent?: string): void => {
+      _registerTheme<ITheme, IThemeDefinition>(graph, definition, name, parent);
+    },
+    getTheme: (name?: string): ITheme => {
+      return _getTheme<ITheme, IThemeDefinition>(graph, resolver, name);
+    },
+    updatePlatformDefaults: (newDefaults: ITheme): void => {
+      _updatePlatformDefaults<ITheme, IThemeDefinition>(graph, newDefaults);
+    }
+  };
+}
+
+/**
+ * Register a new theme by name.  This will invalidate the portion of the graph that is dependent on the new
+ * theme.  This will error on bad dependencies including cycles.
+ *
+ * @param graph - theme graph used for tracking dependencies
+ * @param definition - new theme definition or processing function to set in the theme
+ * @param name - name of the theme, if empty will be treated as 'default'
+ * @param parent - optional theme parent, if empty will be treated as 'default' if not default theme
+ */
+function _registerTheme<ITheme, IThemeDefinition>(
+  graph: IThemeGraph<ITheme, IThemeDefinition>,
+  definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>,
+  name?: string,
+  parent?: string
+): void {
+  name = name || _defaultKey;
+  parent = (name === _defaultKey && _platformKey) || parent || _defaultKey;
+  if (_wouldCauseCycle(graph, name, parent)) {
+    throw new Error('Attempt to register a dependent theme that would cause a cycle');
+  } else if (!graph[parent]) {
+    throw new Error('Attempting to parent to an unknown theme');
+  } else if (name === _platformKey) {
+    throw new Error('A platform theme cannot be registered, call updatePlatformDefaults instead');
+  }
+  if (!graph[name]) {
+    graph[name] = {};
+  }
+  const entry = graph[name];
+  entry.parent = parent;
+  entry.definition = definition;
+  entry.resolved = undefined;
+  _clearDependentThemes(graph, name);
+}
+
+/**
+ * Get a cached theme (if valid) or generate the theme and return it if not
+ *
+ * @param graph - theme graph used for storage and lookups
+ * @param resolver - resolution function used to process themes
+ * @param name - name of the theme to query
+ */
+function _getTheme<ITheme, IThemeDefinition>(
+  graph: IThemeGraph<ITheme, IThemeDefinition>,
+  resolver: ResolveTheme<ITheme, IThemeDefinition>,
+  name?: string
+): ITheme {
+  name = name || _defaultKey;
+  const entry = graph[name] || graph.default;
+  if (!entry.resolved) {
+    const parent = _getTheme(graph, resolver, entry.parent);
+    const fnDef = entry.definition && typeof entry.definition === 'function' && entry.definition;
+    const definition = fnDef ? (fnDef as ProcessTheme<ITheme, IThemeDefinition>)(parent) : entry.definition as IThemeDefinition;
```

Comment:
Hmm, I'm thinking neither of these casts should be needed here.. TS should be able to recognize the `typeof` filter above. Casting in general should be avoided if possible to prevent hiding typing issues. Are you getting an error if it's removed? If so, should be able to restructure this to remove the cast. [Here](https://github.com/OfficeDev/office-ui-fabric-react/blob/73c0e178e3a920d46987ca781cedf7abcbd2c729/packages/experiments/src/utilities/slots.tsx#L94) is an example of an arg that can be an object or function being used without casting. Maybe the best way to do that is just with an `if`/`else` block on `typeof entry.definition`.